src/DynamicContent/Strings.php

```diff
@@ -28,41 +34,112 @@ class Strings {
 	 */
 	public static function filter( array $strings, $nodeId, array $element ) {
 
-		$dynamicFields = wpml_collect( isset( $element['settings']['__dynamic__'] ) ? $element['settings']['__dynamic__'] : [] );
+		$dynamicFields = self::getDynamicFields( $element );
 
-		$getStringKey = function ( WPML_PB_String $string ) {
-			return explode( '-', $string->get_name() )[0];
-		};
+		$updateFromDynamicFields = function( WPML_PB_String $string ) use ( &$dynamicFields, $nodeId ) {
+			$matchingField = $dynamicFields->first( function( Field $field, $key ) use ( $string ) {
+				return $field->isMatchingStaticString( $string );
+			} );
 
-		$updateFromDynamicFields = function( WPML_PB_String $string, $fieldKey ) use ( &$dynamicFields, $nodeId ) {
-			if ( $dynamicFields->has( $fieldKey ) ) {
-				return self::addBeforeAfterAndFallback( wpml_collect( [ $fieldKey => $dynamicFields->pull( $fieldKey ) ] ), $nodeId );
+			if ( $matchingField ) {
+				return self::addBeforeAfterAndFallback( wpml_collect( [ $dynamicFields->pull( $dynamicFields->search( $matchingField ) ) ] ), $nodeId );
 			}
 
 			return $string;
 		};
 
 		return wpml_collect( $strings )
-			->keyBy( $getStringKey )
 			->map( $updateFromDynamicFields )
 			->merge( self::addBeforeAfterAndFallback( $dynamicFields, $nodeId ) )
 			->flatten()
 			->toArray();
 	}
 
+	/**
+	 * @param array $element
+	 *
+	 * @return Collection
+	 */
+	private static function getDynamicFields( array $element ) {
+		if ( self::isModuleWithItems( $element ) ) {
+			return self::getDynamicFieldsForModuleWithItems( $element );
+		} elseif ( isset( $element[ self::KEY_SETTINGS ][ self::KEY_DYNAMIC ] ) ) {
+			return self::getFields(
+				$element[ self::KEY_SETTINGS ][ self::KEY_DYNAMIC ],
+				$element[ self::KEY_NODE_ID ]
+			);
+		}
+
+		return wpml_collect();
+	}
+
+	/**
+	 * @param array $element
+	 *
+	 * @return Collection
+	 */
+	private static function getDynamicFieldsForModuleWithItems( array $element ) {
+		return wpml_collect( reset( $element[ self::KEY_SETTINGS ] ) )
+			->filter( function( array $item ) {
+				return isset( $item[ self::KEY_DYNAMIC ] );
+			} )
+			->map( function( array $item ) use ( $element ) {
+				return self::getFields(
+					$item[ self::KEY_DYNAMIC ],
+					$element[ self::KEY_NODE_ID ],
+					$item[ self::KEY_ITEM_ID ]
+				);
+			} )
+			->flatten();
```

Comment:
@strategio In general I think it's easier to understand if the functions are extracted out:

```php
$isDynamic = function( array $item ) { return isset( $item[ self::KEY_DYNAMIC ] ); };
$getFields = function( array $item ) use ( $element ) {
				return self::getFields(
					$item[ self::KEY_DYNAMIC ],
					$element[ self::KEY_NODE_ID ],
					$item[ self::KEY_ITEM_ID ]
				);
			};

return wpml_collect( reset( $element[ self::KEY_SETTINGS ] ) )
			->filter( $isDynamic )
			->map( $getFields )
			->flatten();
```

I think it should be done here and elsewhere in this MR.

Reply:
Done @brucepearson! I also prefer to have named closures.